test/model/typing_status_test.dart

```diff
@@ -210,4 +248,247 @@ void main() {
       checkNotifiedOnce();
     }));
   });
+
+  group('handle set typing status', () {
+    late PerAccountStore store;
+    late TypingNotifier model;
+    late FakeApiConnection connection;
+    late TopicNarrow narrow;
+
+    void sendTypingNotification(FakeAsync async, SendableNarrow? narrow, {
+      int expectedRequestCount = 1,
+    }) {
+      for (int i = 0; i < expectedRequestCount; i++) {
+        connection.prepare(json: {});
+      }
+      if (narrow == null) {
+        model.stoppedComposing();
+      } else {
+        model.keystroke(narrow);
+      }
+      async.elapse(Duration.zero);
+    }
+
+    void checkTypingRequest(TypingOp op, SendableNarrow narrow) =>
+      checkSetTypingStatusRequests(connection, [(op, narrow)]);
+
+    Future<void> prepare() async {
+      addTearDown(testBinding.reset);
+      // Only test with the latest behavior of the setTypingStatus API.
+      final account = eg.account(
+        user: eg.selfUser, zulipFeatureLevel: eg.futureZulipFeatureLevel);
+      store = eg.store(account: account, initialSnapshot: eg.initialSnapshot(
+        zulipFeatureLevel: eg.futureZulipFeatureLevel));
+      model = store.typingNotifier;
+      connection = store.connection as FakeApiConnection;
+
+      final channel = eg.stream();
+      await store.addStream(channel);
+      await store.addSubscription(eg.subscription(channel));
+      narrow = TopicNarrow(channel.streamId, 'topic');
+    }
+
+    /// Prepares store and triggers a "typing started" notice.
+    Future<void> prepareStartTyping(FakeAsync async) async {
+      await prepare();
+      sendTypingNotification(async, narrow);
+      checkTypingRequest(TypingOp.start, narrow);
+      check(async.pendingTimers).single;
+    }
+
+    /// Ensures that a "typing stopped" notice is sent when the test ends.
+    void checkTrailingTypingStoppedRequest(SendableNarrow narrow) {
+      connection.prepare(json: {});
+      addTearDown(() => checkTypingRequest(TypingOp.stop, narrow));
+    }
+
+    test('start typing repeatedly extends idle timer', () => awaitFakeAsync((async) async {
+      // t = 0: Start typing. The idle timer is set to typingStoppedWaitPeriod.
+      await prepareStartTyping(async);
+
+      const waitTime = Duration(milliseconds: 100);
+      // [waitTime] should not be long enough
+      // to trigger a "typing stopped" notice.
+      assert(waitTime < model.typingStoppedWaitPeriod);
+
+      async.elapse(waitTime);
+      // t = 100: The idle timer is reset to typingStoppedWaitPeriod.
+      sendTypingNotification(async, narrow);
```

Comment:
This request [doesn't get sent](https://github.com/zulip/zulip-flutter/pull/934#discussion_r1796303132) until the `async.elapse` call after `model.keystroke`, because we `unawaited` the `setTypingStatus` call. We should probably reintroduce futures to `keystroke` et al. so that the callers can `await` them.

Reply:
Hmm. That linked discussion is in a pretty different context: it was a widget test, which manages async differently from `awaitFakeAsync`. And in that situation, returning a future and trying to await it is precisely the thing that *doesn't* work at all, and instead gets stuck.

As described in my other couple of comments just now, it looks like we don't want the callers in the actual app to await this. Which makes sense because we really don't care about the results of this API request — there's nothing we'll do with the response, and if it fails then we'll just shrug because it's inherently transient information anyway.

Given that the actual app doesn't want to await this, there is still an option to return a future just for the sake of tests, but it'd be better to avoid that if we can. In general, it's best to where possible keep any given piece of internal API the way the consumers in the actual app want it, and not complicate it for the sake of tests except where necessary.

So if an `async.elapse()` call is what it takes to make the tests work fine with the simpler API, then let's go for that.

(I'm still a bit puzzled that even that much is necessary, because it looks like the implementation goes ahead and *makes* the request synchronously — the only thing that happens asynchronously is for the request to finish.)

Reply:
I found that what actually happens is that the `async.elapse` is only needed when we check the timers in some cases. Both `new Timer` and `setTypingStatus` creates a timer tracked by `async.pendingTimers`, so whenever we want to see if `new Timer` works, or that `stoppedComposing` cancels the timer, there will also be a pending request.

Reply:
I don't totally follow. What's an example of when we want to see if `new Timer` works? (Surely it always works, right? And if it somehow doesn't, that's an upstream problem.)

Reply:
Yeah, `new Timer` itself works. We want to test that `keystroke` results in a timer being created (or extended), and that `stoppedComposing` cancels the timer. Both methods might send an API request that also counts toward `async.pendingTimers` with `Duration.zero`.